_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,753 @@
+| To achieve this goal:                                                             | in `torch` do:    | in Keras do:                                   |
++===================================================================================+===================+================================================+
+| access all intermediate outputs ( = per-time-step outputs)                        | `ret[[1]]`        | `return_sequences = TRUE`                      |
++-----------------------------------------------------------------------------------+-------------------+------------------------------------------------+
+| access both "hidden state" (output) and "cell state" from final time step (only!) | `ret[[2]]`        | `return_state = TRUE`                          |
++-----------------------------------------------------------------------------------+-------------------+------------------------------------------------+
+| access all intermediate outputs and the final "cell state"                        | both of the above | `return_sequences = TRUE, return_state = TRUE` |
++-----------------------------------------------------------------------------------+-------------------+------------------------------------------------+
+| access all intermediate outputs and "cell states" from all time steps             | no way            | no way                                         |
++-----------------------------------------------------------------------------------+-------------------+------------------------------------------------+
+
+: Table 2: Contrasting ways of obtaining various return values in `torch` vs. Keras. Cf. the appendix for complete examples.
+:::
+
+<br />
+
+### `convLSTM`, the plan
+
+In both `torch` and Keras RNN architectures, single time steps are processed by corresponding `Cell` classes: There is an LSTM Cell matching the LSTM, a GRU Cell matching the GRU, and so on. We do the same for ConvLSTM. In `convlstm_cell()`, we first define what should happen to a single observation; then in `convlstm()`, we build up the recurrence logic.
+
+Once we're done, we create a dummy dataset, as reduced-to-the-essential as can be. With more complex datasets, even artificial ones, chances are that if we don't see any training progress, there are hundreds of possible explanations. We want a sanity check that, if we fail it, leaves no excuses. Realistic applications are left to future posts.
+
+### A single step: `convlstm_cell`
+
+Our `convlstm_cell`'s constructor takes arguments `input_dim` , `hidden_dim`, and `bias`, just like a `torch` LSTM Cell.
+
+But we're processing two-dimensional input data. Instead of the usual affine combination of new input and previous state, we use a convolution, of kernel size `kernel_size`. Inside `convlstm_cell`, it is `self$conv` that takes care of this.
```

Comment:
```suggestion
But we're processing two-dimensional input data. Instead of the usual affine combination of new input and previous state, we use a convolution of kernel size `kernel_size`. Inside `convlstm_cell`, it is `self$conv` that takes care of this.
```